From 8adba35697e27b62e5147e3b9b5b244e4d97e1ca Mon Sep 17 00:00:00 2001
From: Adarsh-MR <adarshmr1998@gmail.com>
Date: Sun, 27 May 2018 08:04:03 +0000
Subject: [PATCH 36/37] Themes: base: expose QS tiles and label colors

* also exposed expand indicator and underline colors

* also disabling grayscale tint if config_enable_qs_tile_tinting enabled.
  because switching tiles state (enable/disable) will apply grayscale tint and removes
  themed tint color,  big thanks to Gurpreet ( SNN Infinity) for headsup on this.

  this bool also prevents black on black issue if a themer doesn't change tiles tint colors.

Change-Id: Iba15ad4013378a815607894a36ee18ea30bfcc3b
Signed-off-by: Adarsh-MR <adarshmr1998@gmail.com>
---
 packages/SystemUI/res/layout/qs_tile_label.xml  |  8 ++++----
 packages/SystemUI/res/values/projekt_colors.xml |  9 +++++++++
 packages/SystemUI/res/values/projekt_config.xml |  4 ++++
 .../systemui/qs/tileimpl/QSIconViewImpl.java    | 12 ++++++++++--
 .../systemui/qs/tileimpl/QSTileImpl.java        | 17 +++++++++++++++--
 .../res/values/projekt_colors.xml               |  5 +++++
 6 files changed, 47 insertions(+), 8 deletions(-)

diff --git a/packages/SystemUI/res/layout/qs_tile_label.xml b/packages/SystemUI/res/layout/qs_tile_label.xml
index b138df08625..468dfb3b5bc 100644
--- a/packages/SystemUI/res/layout/qs_tile_label.xml
+++ b/packages/SystemUI/res/layout/qs_tile_label.xml
@@ -46,7 +46,7 @@
             android:gravity="center"
             android:ellipsize="end"
             android:textAppearance="@style/TextAppearance.QS.TileLabel"
-            android:textColor="?android:attr/textColorPrimary"/>
+            android:textColor="@color/qs_tiles_label_color"/>
 
         <ImageView android:id="@+id/restricted_padlock"
             android:layout_width="@dimen/qs_tile_text_size"
@@ -63,7 +63,7 @@
             android:layout_width="18dp"
             android:layout_height="match_parent"
             android:src="@drawable/qs_dual_tile_caret"
-            android:tint="?android:attr/textColorPrimary"
+            android:tint="@color/qs_tiles_expand_tint"
             android:visibility="gone" />
     </LinearLayout>
 
@@ -80,7 +80,7 @@
         android:visibility="gone"
         android:gravity="center"
         android:textAppearance="@style/TextAppearance.QS.TileLabel"
-        android:textColor="?android:attr/textColorPrimary"/>
+        android:textColor="@color/qs_tiles_label_color"/>
 
     <View
         android:id="@+id/underline"
@@ -92,6 +92,6 @@
         android:layout_below="@id/label_group"
         android:visibility="gone"
         android:alpha="?android:attr/disabledAlpha"
-        android:background="?android:attr/colorForeground"/>
+        android:background="@color/qs_tiles_underline_color"/>
 
 </com.android.systemui.qs.tileimpl.ButtonRelativeLayout>
diff --git a/packages/SystemUI/res/values/projekt_colors.xml b/packages/SystemUI/res/values/projekt_colors.xml
index a6b658c0514..2bd8733f532 100644
--- a/packages/SystemUI/res/values/projekt_colors.xml
+++ b/packages/SystemUI/res/values/projekt_colors.xml
@@ -39,4 +39,13 @@
     <color name="screen_pinning_positive_text_color">@android:color/white</color>
     <color name="screen_pinning_negative_text_color">@android:color/white</color>
 
+    <!-- QS tiles Active and Inactive tint (Requires config_enable_qs_tile_tinting bool set to true) -->
+    <color name="qs_tiles_active_tint">@*android:color/primary_text_default_material_light</color>
+    <color name="qs_tiles_inactive_tint">@*android:color/hint_foreground_material_light</color>
+    
+    <!-- QS tiles expand, underline and label colors -->    
+    <color name="qs_tiles_expand_tint">@*android:color/text_color_primary</color>
+    <color name="qs_tiles_label_color">@*android:color/text_color_primary</color>
+    <color name="qs_tiles_underline_color">@*android:color/foreground_material_light</color>
+
 </resources>
diff --git a/packages/SystemUI/res/values/projekt_config.xml b/packages/SystemUI/res/values/projekt_config.xml
index 3833555f971..9b2aedf3cbf 100644
--- a/packages/SystemUI/res/values/projekt_config.xml
+++ b/packages/SystemUI/res/values/projekt_config.xml
@@ -15,4 +15,8 @@
 -->
 <resources>
     <item name="qs_background_opacity" format="float" type="integer">0.87</item>
+
+    <!-- Config to enable/diaable qs tile tinting -->
+    <bool name="config_enable_qs_tile_tinting">false</bool>
+
 </resources>
diff --git a/packages/SystemUI/src/com/android/systemui/qs/tileimpl/QSIconViewImpl.java b/packages/SystemUI/src/com/android/systemui/qs/tileimpl/QSIconViewImpl.java
index c249e3778c0..491910b4f21 100644
--- a/packages/SystemUI/src/com/android/systemui/qs/tileimpl/QSIconViewImpl.java
+++ b/packages/SystemUI/src/com/android/systemui/qs/tileimpl/QSIconViewImpl.java
@@ -155,17 +155,21 @@ public class QSIconViewImpl extends QSIconView {
         return getColorForState(getContext(), state);
     }
 
-    public static void animateGrayScale(int fromColor, int toColor, ImageView iv) {
+    public void animateGrayScale(int fromColor, int toColor, ImageView iv) {
+
         if (iv instanceof AlphaControlledSlashImageView) {
             ((AlphaControlledSlashImageView)iv)
                     .setFinalImageTintList(ColorStateList.valueOf(toColor));
         }
+
         if (ValueAnimator.areAnimatorsEnabled()) {
             final float fromAlpha = Color.alpha(fromColor);
             final float toAlpha = Color.alpha(toColor);
             final float fromChannel = Color.red(fromColor);
             final float toChannel = Color.red(toColor);
 
+            boolean enableQsTileTinting = getContext().getResources().getBoolean(R.bool.config_enable_qs_tile_tinting);
+
             ValueAnimator anim = ValueAnimator.ofFloat(0, 1);
             anim.setDuration(QS_ANIM_LENGTH);
             anim.addUpdateListener(animation -> {
@@ -173,7 +177,11 @@ public class QSIconViewImpl extends QSIconView {
                 int alpha = (int) (fromAlpha + (toAlpha - fromAlpha) * fraction);
                 int channel = (int) (fromChannel + (toChannel - fromChannel) * fraction);
 
-                setTint(iv, Color.argb(alpha, channel, channel, channel));
+                if (!enableQsTileTinting) {
+                    setTint(iv, Color.argb(alpha, channel, channel, channel));
+                } else {
+                    setTint(iv, toColor);
+                }
             });
 
             anim.start();
diff --git a/packages/SystemUI/src/com/android/systemui/qs/tileimpl/QSTileImpl.java b/packages/SystemUI/src/com/android/systemui/qs/tileimpl/QSTileImpl.java
index 576a447447b..a3d9faaa956 100644
--- a/packages/SystemUI/src/com/android/systemui/qs/tileimpl/QSTileImpl.java
+++ b/packages/SystemUI/src/com/android/systemui/qs/tileimpl/QSTileImpl.java
@@ -50,6 +50,8 @@ import com.android.systemui.plugins.qs.QSTile.State;
 import com.android.systemui.qs.PagedTileLayout.TilePage;
 import com.android.systemui.qs.QSHost;
 
+import com.android.systemui.R;
+
 import java.util.ArrayList;
 
 /**
@@ -370,14 +372,25 @@ public abstract class QSTileImpl<TState extends State> implements QSTile {
     public abstract CharSequence getTileLabel();
 
     public static int getColorForState(Context context, int state) {
+        
+        boolean enableQsTileTinting = context.getResources().getBoolean(R.bool.config_enable_qs_tile_tinting);
+
         switch (state) {
             case Tile.STATE_UNAVAILABLE:
                 return Utils.getDisabled(context,
                         Utils.getColorAttr(context, android.R.attr.colorForeground));
             case Tile.STATE_INACTIVE:
-                return Utils.getColorAttr(context, android.R.attr.textColorHint);
+                if (!enableQsTileTinting) {
+                    return Utils.getColorAttr(context, android.R.attr.textColorHint);
+                } else {
+                    return context.getResources().getColor(R.color.qs_tiles_inactive_tint);
+                }
             case Tile.STATE_ACTIVE:
-                return Utils.getColorAttr(context, android.R.attr.textColorPrimary);
+                if (!enableQsTileTinting) {
+                    return Utils.getColorAttr(context, android.R.attr.textColorPrimary); 
+                } else {
+                    return context.getResources().getColor(R.color.qs_tiles_active_tint);
+                }
             default:
                 Log.e("QSTile", "Invalid state " + state);
                 return 0;
diff --git a/packages/overlays/SysuiDarkThemeOverlay/res/values/projekt_colors.xml b/packages/overlays/SysuiDarkThemeOverlay/res/values/projekt_colors.xml
index 354bb20dba3..1aafc76d23e 100644
--- a/packages/overlays/SysuiDarkThemeOverlay/res/values/projekt_colors.xml
+++ b/packages/overlays/SysuiDarkThemeOverlay/res/values/projekt_colors.xml
@@ -20,4 +20,9 @@
     <!-- QS edit page background color -->
     <color name="qs_edit_item_decoration_bg">@*android:color/secondary_device_default_settings</color>
 
+    <!-- QS tiles and label colors -->
+    <color name="qs_tiles_active_tint">@*android:color/primary_text_default_material_dark</color>
+    <color name="qs_tiles_inactive_tint">@*android:color/hint_foreground_material_dark</color>
+    <color name="qs_tiles_underline_color">@*android:color/foreground_material_dark</color>
+
 </resources>
-- 
2.18.0

